### DISCUSSION TITLE
Ability to control number of concurrent builds from revalidate to avoid server overload/crash

### DISCUSSION THREAD

**Discussion by karlmacklin** [vercel/next.js#41492] [Ideas]
### Describe the feature you'd like to request

Problem scenario:

Website for a given customer has ~1500 pages. These pages can be built via getStaticPaths/getStaticProps with `npm run build`. Build works fine and nextjs builds the pages in an orderly fashion, not overloading the system by not trying to build everything at the same time.

Fast forward to a web crawler fetching the sitemap and hitting all the pages in a super short amount of time.
An individual getStaticprops call is not overly costly, but hitting them all at once with no queue system at all causes the server to overload.

What is happening?
Well, all the pages are pre-rendered and served to the crawler extremely fast, as they should. However, they are all stale caches, so via the revalidate property we are telling nextjs to first serve the stale site and in the background render a new version.

**There does not seem to be a limit how many concurrent builds nextjs will attempt to do background revalidate renders, in contrast to the actual build process where the concurrency is far more reasonable.**

So we need a way to tone this down, or else the client (and their server specifically) will remain unhappy.

### Describe the solution you'd like

I can see two possible ways to solve this:

1. A config setting that controls how many concurrent builds nextjs will attempt at a time and put the rest of the requests in a queue. For any new request made while the queue is full enough to go over the limit, allow another option to control whether or not the user should be served the current stale (if available), or if they will have to wait.

2. A way to access the nextjs cache status from within a middleware so we at least can build our queue system. I feel nextjs developers should probably not have to deal with that question in a perfect world but this could perhaps be a powerful tool for more use cases.
Access to the nextjs cache status would be needed to determine how many of the requests are causing a new build. We would preferrably also have access to status about current builds in progress.

### Describe alternatives you've considered

Adding an extra load balancer in front to introduce an artifical queue, or adding ip based rate limiting however some crawlers even employ multi ip solutions so that is not a surefire way to deal with it.

**Comment by balazsorban44** [comment]
Hi, instead of a time-based revalidation, have you considered On-demand Revalidation?

See the docs:
https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation

Basically, instead of letting a crawler revalidate all your pages at once, make the revalidation when the CMS/data source changes instead, per page. So the crawler not only will see the newest version instantly, but the server won't revalidate multiple pages at once.

**Reply by karlmacklin** [reply]
Hi, sorry I realize I was a bit vague in my original post.

The crawler is not ours to control - think more as an SEO spider that is a bit too eager on its concurrency.

We do use a combination of time based and on demand generation.

When the CMS publishes a new version it calls revalidate on that slug, it works great.
However, the CMS controls more global things as well, like main menu/global menu, related articles etc. All of these can change and thus it is also required for us to keep a maximum lifetime of pages so they get a chance to eventually catch up.
